whereby/sdk
# common: Add notifications eventing system

### Description

Add a general notification system for push-based events in the SDK, implement a selection of initial notification events firing and provide some examples of notification event handling in both the storybook and sample apps.

To subscribe to notification events a user can add the following code:

```javascript
const roomConnection = useRoomConnection("__room_url__")

const wherebyNotificationEventHandler = (event) => console.log("received notification event", event);

// Use the wildcard "*" character to listen for ALL notification events
roomConnection.events.on("*", wherebyNotificationEventHandler);

// Alternatively you can subscribe to individual events as follows:
// roomConnection.events.on("chatMessageReceived", (event) => console.log("chat message received", event));

// ...

// Unsubscribe from notification events at some future point (on e.g. unmount of component)
React.useEffect(() => () => roomConnection.events.off("*", wherebyNotificationEventHandler), []);
```

The initial notification events implemented in this PR include the following:

* `remoteHandRaised`: a remote participant has raised their hand to speak in the meeting
* `remoteHandLowered`: a remote participant has lowered their hand to speak in the meeting
* `requestAudioEnable`: a meeting host has asked you to speak
* `requestAudioDisable`: a meeting host has remotely muted your microphone
* `signalTrouble`: the connection to the signalling server has been interrupted due to network issues
* `signalOk`: the connection to the signalling server should be working normally
* `chatMessageReceived`: a chat message was received from a remote participant

### Testing

1. Check out this branch and run `yarn install` followed by `yarn build && yarn dev`
2. Once storybook is up an running open the "Room Connection With Host Controls" story (direct link when storybook is running [here](http://localhost:6006/?path=/story/examples-custom-ui--room-connection-with-host-controls)).
3. Join the same meeting room directly by its URL (not using the SDK)
4. Check that all "raise hand" functionality works correctly (both raising the local participants hand + asking a remote participant to speak after they raise their hand in the room also).
5. Check all other notifications are firing as expected.

### Checklist

        under the terms of the MIT license.
